src/zkevm_specs/evm/execution/oog_constant.py

```diff
@@ -0,0 +1,44 @@
+from ...util import FQ
+from ..instruction import Instruction, Transition
+from ..table import CallContextFieldTag
+from ..execution_state import ExecutionState
+from ..opcode import Opcode
+from ...util import N_BYTES_GAS
+
+
+def oog_constant(instruction: Instruction):
+    # retrieve op code associated to oog constant error
+    opcode = instruction.curr.aux_data
+
+    # check gas left is less than const gas required
+    gas_not_enough, _ = instruction.compare(
+        instruction.curr.gas_left, FQ(Opcode(opcode.expr().n).constant_gas_cost()), N_BYTES_GAS
+    )
+    instruction.constrain_equal(gas_not_enough, FQ(1))
+
+    # current call must be failed.
+    is_success = instruction.call_context_lookup(CallContextFieldTag.IsSuccess)
+    instruction.constrain_equal(is_success, FQ(0))
+
+    # Go to EndTx only when is_root
+    is_to_end_tx = instruction.is_equal(instruction.next.execution_state, ExecutionState.EndTx)
+    instruction.constrain_equal(FQ(instruction.curr.is_root), is_to_end_tx)
+
+    if instruction.curr.is_root:
+        # When a transaction ends with this error state, this call must not be persistent
+        is_persistent = instruction.call_context_lookup(CallContextFieldTag.IsPersistent)
```

Comment:
I think all calls that are `IsSuccess = false` are also `IsPersistent = false`, not just in this particular case.  So in other words, the case `IsSuccess = false and IsPersistent = true` is invalid.
Right now I'm not sure if we're checking this, but I think it should be checked for every step and outside of the gadget definition.

Reply:
Only `false-false` and `true-true` should be valid to be witnessed no??? If so, we can constraint them to be equal. Making sure there's no way arround it to trick them.

Reply:
when  `IsSuccess` = true, `IsPersistent` can be false since caller's `IsSuccess` can be false. these two aren't always equally . but when` IsSuccess = false`,  `IsPersistent` must be `false`, in error cases , they are identical. I will remove root call condition, and check like :
  ```
    is_success = instruction.call_context_lookup(CallContextFieldTag.IsSuccess)
    instruction.constrain_equal(is_success, FQ(0))
    is_persistent = instruction.call_context_lookup(CallContextFieldTag.IsPersistent)
    instruction.constrain_equal(is_success, is_persistent)
```
let me know if you think differently!

Reply:
> when `IsSuccess` = true, `IsPersistent` can be false since caller's `IsSuccess` can be false. these two aren't always equally . but when` IsSuccess = false`, `IsPersistent` must be `false`, in error cases , they are identical. I will remove root call condition, and check like :
> 
> ```
>   is_success = instruction.call_context_lookup(CallContextFieldTag.IsSuccess)
>   instruction.constrain_equal(is_success, FQ(0))
>   is_persistent = instruction.call_context_lookup(CallContextFieldTag.IsPersistent)
>   instruction.constrain_equal(is_success, is_persistent)
> ```
> 
> we can refactor to common helper or gadget when implement second error state as all error states share this constraint. let me know if you think differently!

This approach sounds good to me!  The second constraint could also be:
 `instruction.constrain_equal(is_persistent, FQ(0))`, which is equivalent (and maybe simpler)